python/packages/autogen-ext/src/autogen_ext/code_executors/jupyter/_jupyter_code_executor.py

```diff
@@ -0,0 +1,151 @@
+import asyncio
+import base64
+import json
+import sys
+import uuid
+from dataclasses import dataclass
+from pathlib import Path
+from types import TracebackType
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+
+from autogen_core import CancellationToken
+from autogen_core.code_executor import CodeBlock, CodeExecutor, CodeResult
+
+from .._common import silence_pip
+from ._jupyter_connectable import JupyterConnectable
+
+
+@dataclass
+class JupyterCodeResult(CodeResult):
+    """A code result class for Jupyter code executor."""
+
+    output_files: list[Path]
+
+
+class JupyterCodeExecutor(CodeExecutor):
+    def __init__(
+        self,
+        server: JupyterConnectable,
+        kernel_name: str = "python3",
+        timeout: int = 60,
+        output_dir: Path | str = Path("."),
+    ):
+        """A code executor class that executes code statefully using
+        a Jupyter server supplied to this class.
+
+        Each execution is stateful and can access variables created from previous
+        executions in the same session.
+
+        Args:
+            server (JupyterConnectable): The Jupyter server to use.
+            timeout (int): The timeout for code execution, by default 60.
+            kernel_name (str): The kernel name to use. Make sure it is installed.
+                By default, it is "python3".
+            output_dir (str): The directory to save output files, by default ".".
+        """
+        if timeout < 1:
+            raise ValueError("Timeout must be greater than or equal to 1.")
+
+        if isinstance(output_dir, str):
+            output_dir = Path(output_dir)
+
+        self._jupyter_client = server.get_client()
+        available_kernels = self._jupyter_client.list_kernel_specs()
+        if kernel_name not in available_kernels["kernelspecs"]:
+            raise ValueError(f"Kernel {kernel_name} is not installed.")
+
+        self._kernel_id = self._jupyter_client.start_kernel(kernel_name)
+        self._jupyter_kernel_client = self._jupyter_client.get_kernel_client(self._kernel_id)
+        self._timeout = timeout
+        self._output_dir = output_dir
+
+    async def init(self):
+        # TODO: How can the design be done differently, so this init method is not needed
+        # It seems difficult to me, because the constructor is always sync code
+        await self._jupyter_kernel_client.__aenter__()
```

Comment:
I could need some input here on how to improve this, so this init function is not needed

Reply:
Perhaps we could put this code in execute_code_blocks? I think reopening the socket every once in a while might make sense anyway? This might resolve this bug as well: https://github.com/microsoft/autogen/issues/3477

Reply:
Fixed with new version